### PR TITLE
[CustomOp] Support duplicable op input and output

### PR DESCRIPTION
**Pull request by chenwhql**
### PR types

New features
### PR changes

OPs
### Describe

[CustomOp] Support duplicable op input and output.

This PR add variable length tensor input and output in op, for example:

```c++
// add new param type: const std::vector<paddle::Tensor>&
std::vector<paddle::Tensor> ConcatForwardDynamicAxis(
    const std::vector<paddle::Tensor>& inputs, const paddle::Tensor& axis_t) {
  ...
}

// add new param type: std::vector<std::vector<int64_t>>
std::vector<std::vector<int64_t>> ConcatInferShapeDynamicAxis(
    std::vector<std::vector<int64_t>> input_shapes,
    std::vector<int64_t> axis_shape) {
  return {std::vector<int64_t>(input_shapes[0].size(), -1)};
}

// add new param type: std::vector<paddle::DataType
std::vector<paddle::DataType> ConcatInferDtypeDynamicAxis(
    std::vector<paddle::DataType> input_dtypes, paddle::DataType axis_dtype) {
  return {input_dtypes[0]};
}

// the input string of `const std::vector<paddle::Tensor>&` need wrapped by `paddle::Vec()`
PD_BUILD_OP(custom_concat)
    .Inputs({paddle::Vec("X"), "Axis"})
    .Outputs({"Out"})
    .SetKernelFn(PD_KERNEL(ConcatForwardDynamicAxis))
    .SetInferShapeFn(PD_INFER_SHAPE(ConcatInferShapeDynamicAxis))
    .SetInferDtypeFn(PD_INFER_DTYPE(ConcatInferDtypeDynamicAxis));

// note the warp order: paddle::Grad(paddle::Vec("X"))
PD_BUILD_GRAD_OP(custom_concat)
    .Inputs({paddle::Vec("X"), paddle::Grad("Out"), "Axis"})
    .Outputs({paddle::Grad(paddle::Vec("X"))})
    .SetKernelFn(PD_KERNEL(ConcatBackwardDynamicAxis));
```

### CODE REVIEW

File: paddle/fluid/extension/include/ext_op_meta_info.h [hunk 1]
```diff
@@ -111,15 +128,32 @@ struct KernelFuncImpl<Return (*)(Args...), impl_fn> {
...
       const Tensor& arg = inputs[in_idx];
-      return ComputeCallHelper<Tail...>::template Compute<in_idx + 1, attr_idx>(
-          inputs, attrs, pargs..., arg);
+      return ComputeCallHelper<Tail...>::template Compute<in_idx + 1,
+                                                          vec_in_idx, attr_idx>(
+          inputs, vec_inputs, attrs, pargs..., arg);
+    }
+  };
+
+  // for std::vector<Tensor> input
```
Comment (JiabinYang) [inline paddle/fluid/extension/include/ext_op_meta_info.h:144]: Bad comments, describe what's for in detail

File: paddle/fluid/framework/custom_operator.cc [hunk 2]
```diff
@@ -103,19 +107,47 @@ static void RunKernelFunc(const framework::ExecutionContext& ctx,
...
-    CustomTensorUtils::SetTensorCurrentStream(&custom_in, ctx.GetPlace());
-    custom_ins.emplace_back(custom_in);
+    if (detail::IsDuplicableVar(in_name)) {
+      // return const std::vector<const Tensor*>
+      auto vec_x = ctx.MultiInput<Tensor>(in_name);
+      PADDLE_ENFORCE_NE(vec_x.empty(), true,
+                        platform::errors::NotFound(
+                            "Input vector<tensor> (%s) is empty.", in_name));
+      std::vector<paddle::Tensor> custom_vec_in;
+      for (size_t i = 0; i < vec_x.size(); ++i) {
```
Comment (JiabinYang) [inline paddle/fluid/framework/custom_operator.cc:120]: clang format using `for(const auto& x: vec_x)` is better?

**Review by JiabinYang — APPROVED** [review]
LGTM

File: paddle/fluid/extension/include/ext_op_meta_info.h (hunk 1 above)
Comment (chenwhql) [reply paddle/fluid/extension/include/ext_op_meta_info.h:144]: thx, I will remove it in next PR

File: paddle/fluid/framework/custom_operator.cc (hunk 2 above)
Comment (chenwhql) [reply paddle/fluid/framework/custom_operator.cc:120]: if using `for(const auto& x: vec_x)`, we cannot tell users which tensor cause error in error message.

File: paddle/fluid/extension/include/ext_op_meta_info.h [hunk 3]
```diff
@@ -155,40 +190,62 @@ struct KernelFuncImpl<Return (*)(Args...), impl_fn> {
...
 
-  // only one type input: std::vector<int64_t>
   template <typename... Tail>
   struct InferShapeCallHelper<std::vector<int64_t>, Tail...> {
-    template <int in_idx, typename... PreviousArgs>
-    static Return InferShape(std::vector<std::vector<int64_t>> input_shapes,
-                             const PreviousArgs&... pargs) {
+    template <int in_idx, int vec_in_idx, typename... PreviousArgs>
+    static Return InferShape(
+        std::vector<std::vector<int64_t>> input_shapes,
```
Comment (Aurelius84) [inline paddle/fluid/extension/include/ext_op_meta_info.h:217]: 我们可以统一一下函数的输入参数类型，infershape、inferDatatype可以都使用`const &`，更符合编程习惯和直觉。可以在下个PR修改

**Review by Aurelius84 — APPROVED** [review]

File: paddle/fluid/extension/include/ext_op_meta_info.h (hunk 3 above)
Comment (chenwhql) [reply paddle/fluid/extension/include/ext_op_meta_info.h:217]: thx，下个PR修改

**Review by XieYunshen — APPROVED** [review]

**Pull request merged**